### PR TITLE
rustc: rename -Z emit-directives to -Z emit-artifact-notifications and simplify the output.

### PR DESCRIPTION
**Pull request by eddyb**
This is my take on #60006 / #60419 (see https://github.com/rust-lang/rust/pull/60006#discussion_r275983732). 
I'm not too attached the "notifications" part, it's pretty much bikeshed material.
**EDIT**: for "artifact", @matklad pointed out Cargo already uses it (in https://github.com/rust-lang/rust/pull/60464#issuecomment-488576998)

The first two commits are fixes that could be landed independently, especially the `compiletest` one, which removes the need for any of the normalization added in #60006 to land the test.

The last commit enables the emission for all outputs, which was my main suggestion for #60006, mostly to show that it's minimal and not really a "scope creep" (as suggested in https://github.com/rust-lang/rust/pull/60006#discussion_r279964081).

cc @alexcrichton @nnethercote

### CODE REVIEW

**Comment by matklad** [comment]
"compiler-artifact" is also [used by cargo](https://github.com/rust-lang/cargo/blob/ba5aac51e149a8ad29cef75b550f28c00f3cf3a4/src/cargo/util/machine_message.rs#L44-L48), so it's an ok name.

**Comment by alexcrichton** [comment]
I don't personally have a preference one way or another on this, so long as Cargo gets the notification for metadata seems reasonable to me!

File: src/tools/compiletest/src/runtest.rs [hunk 1]
```diff
@@ -1422,10 +1422,21 @@ impl<'test> TestCx<'test> {
     }
 
     fn compile_test(&self) -> ProcRes {
-        let mut rustc = self.make_compile_args(
-            &self.testpaths.file,
-            TargetLocation::ThisFile(self.make_exe_name()),
-        );
+        // Only use `make_exe_name` when the test ends up being executed.
```
Comment (alexcrichton) [inline src/tools/compiletest/src/runtest.rs]: Where did these changes come from and/or how are they related to the PR's title?

**Review by alexcrichton — COMMENTED** [review]
Looks reasonable to me! There's some stray changes I don't know why are included though

File: src/tools/compiletest/src/runtest.rs (hunk 1 above)
Comment (eddyb) [reply src/tools/compiletest/src/runtest.rs]: See the PR description - they remove the need for normalization, by fixing the `compiletest` bugs that affected #60006. IMO, the resulting tests are more clearly correct.

Comment (alexcrichton) [reply src/tools/compiletest/src/runtest.rs]: Ah ok, sorry for missing that!

**Comment by eddyb** [comment]
cc @davidtwco (#53196) looks like `$TEST_BUILD_DIR` doesn't work well on windows?

**Comment by davidtwco** [comment]
@eddyb I'll kick off a build of this on Windows and see if I can figure out what it's doing wrong. I probably won't get to that until tomorrow though.

From a quick glance, I notice that the `$TEST_BUILD_DIR` path is printed with `.to_str().unwrap()`:

https://github.com/rust-lang/rust/blob/a19cf18c7dbbcc46dddea81df3a4cee1735c2349/src/tools/compiletest/src/runtest.rs#L3169

Above that, the `$SRC_DIR` is printed with `.display().to_string()`:

https://github.com/rust-lang/rust/blob/a19cf18c7dbbcc46dddea81df3a4cee1735c2349/src/tools/compiletest/src/runtest.rs#L3160

I wonder if this would be as simple as changing the `$TEST_BUILD_DIR` and `$BUILD_DIR` paths to print the same way as `$SRC_DIR`?

**Comment by nnethercote** [comment]
It took me multiple goes to get the normalization right due to various platform-specific differences. I suspect you might get to go through the same experience I had. I suspect you'll have a wasm-related failure at some point.

**Comment by eddyb** [comment]
@nnethercote I fixed the underlying issue (`-o` being passed even when not building a binary to execute), which means I always get the right filename.
And I need no custom normalization, this is just another `compiletest` I want to fix instead of working around it.

**Comment by eddyb** [comment]
Got distracted looking at the history of UI test normalization, I'm thinking https://github.com/rust-lang/rust/pull/53196/commits/0cfc17358aa836440c8b58b364d147ad4ab305de can be partially or fully reverted if the problems in question were caused by this bug.

But in this case, the failure makes a lot of sense, because `if json` logic is missing for most path normalizations, and this is a JSON test, so the replacement from `\` to `\\` is essential.

**Pull request merged**